lib/pipeline_controller.py

```diff
@@ -73,9 +126,99 @@ def process_all_sites_data(self, end_date, batch_num):
             if not self.ignore_cache:
                 self.s3_client.set_cache({"last_poll_date": poll_date.isoformat()})
 
-            self.logger.info(f"Finished batch {batch_num+1}: {poll_date_str}")
+            self.logger.info(f"Finished batch {batch_num+1}: {poll_date.isoformat()}")
             self.process_all_sites_data(end_date, batch_num + 1)
 
+    def process_broken_orbits(self, start_date, end_date):
+        """
+        Re-queries individual sites with unhealthy data from the past 30 days (a limit
+        set by the API) to see if any data has since been recovered
+        """
+        create_table_query = build_redshift_create_table_query(
+            self.redshift_visits_table, start_date, end_date
+        )
+        self.redshift_client.connect()
+        self.redshift_client.execute_transaction([(create_table_query, None)])
+        recoverable_site_dates = self.redshift_client.execute_query(
+            REDSHIFT_RECOVERABLE_QUERY
+        )
+        known_data = self.redshift_client.execute_query(
+            build_redshift_known_query(self.redshift_visits_table)
+        )
+        self.redshift_client.execute_transaction([(REDSHIFT_DROP_QUERY, None)])
```

Comment:
What is the advantage of creating the temporary table here? It seems like it's being joined back to the source table so could this be done with a single query?

Reply:
Yeah I don't love it and went back and forth -- I'd definitely be open to suggestions! The problem is I need to find all sites and dates containing at least one unhealthy orbit, which involves filtering on a few fields and then grouping (`... WHERE NOT is_healthy_data AND is_fresh AND increment_start >= '{start_date}' AND increment_start < '{end_date}' GROUP BY shoppertrak_site_id, increment_date;`), twice: once to get the site/date pairs that I need to query the API for and then again to find the healthy data for those same sites/dates (which is why the temp table gets joined back to the regular table). To avoid creating a temporary table, I have a few options.

I could of course just do the filtering/grouping twice, but that's inefficient. I could also use the values from the first query in an `IN` condition, but that would require either concatenating the site id and date into a single string (`WHERE CONCAT(shoppertrak_site_id, increment_start::DATE) IN (...)`), which feels messy, or doing a multicolumn check (`WHERE (shoppertrak_site_id, increment_start::DATE) IN (SELECT site1, date1 UNION SELECT site1, date1 UNION ...)`, which has atrocious syntax. Finally, I could avoid doing the filtering/grouping on the server and get all of the data at once and then filter/group in Python. That would probably require me to use Pandas, though, which so far the code hasn't required, and is of course space inefficient. Of all these options I chose the temporary table but could be persuaded to do the last one. What do you think?